File: BOJ/[1167] 트리의 지름/KingDonggyu/1167.js
function solution(V, edges) {
  const graph = Array.from(Array(V + 1), () => []);
  edges.forEach((edge) => {
    edge = edge.split(' ').map((x) => +x);
    for (let i = 1; i < edge.length - 1; i += 2) {
      graph[edge[0]].push([edge[i], edge[i + 1]]);
    }
  });

  const [node] = dfs(graph, 1);
  const [_, cost] = dfs(graph, node);
  return cost;
}

function dfs(graph, startNode) {
  let farthestNode = [null, 0];
  const stack = [[startNode, 0, new Set([startNode])]];

  while (stack.length) {
    const [node, cost, visited] = stack.pop();
    if (farthestNode[1] < cost) {
      farthestNode = [node, cost];
    }
    graph[node].forEach(([nextNode, newCost]) => {
      if (visited.has(nextNode)) {
        return;
      }
      const nextVisited = new Set([...Array.from(visited)]);
      nextVisited.add(nextNode);
      stack.push([nextNode, cost + newCost, nextVisited]);
    })
  }

  return farthestNode;
}

const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];

rl.on('line', (line) => {
  input.push(line);
  input.length - 1 === +input[0] && rl.close();
}).on('close', () => {
  console.log(solution(+input[0], input.slice(1)));
  process.exit();
});

File: BOJ/[1167] 트리의 지름/JFe/1167.py
import sys
from collections import defaultdict

def dfs(node, dist):
    global max_dist, max_node
    # 최댓값 업데이트
    if max_dist < dist:
        max_dist = dist
        max_node = node
    visited[node] = True
    # 연결된 다음 노드 탐색
    for next_node, next_dist in graph[node]:
        if visited[next_node]: continue
        next_dist += dist
        dfs(next_node, next_dist)

V = int(sys.stdin.readline())
graph = defaultdict(list)
visited = [False for _ in range(V + 1)]
max_dist, max_node = 0, 0
# 트리 연결
for _ in range(V):
    temp = list(map(int, sys.stdin.readline().split()))
    for i in range(1, len(temp) - 1, 2):
        graph[temp[0]].append([temp[i], temp[i + 1]])
# 임의의 지점 1에서 갈 수 있는 가장 먼 노드 dfs 탐색
dfs(1, 0)
visited = [False for _ in range(V + 1)]
# dfs 탐색 후 나온 가장 먼 노드에서 갈 수 있는 가장 먼 노드 탐색
dfs(max_node, 0)
print(max_dist)